### PR TITLE
support for input controller pose filters

### PR DESCRIPTION
**Pull request by ZappoMan**
Adds support for various filters to run against "pose" style inputs. This will be useful in implementing mo-cap from vive pucks without going through JS update loops.

Test plan:
- smoke test
- make sure hand controllers still work normally
- run [this script](https://raw.githubusercontent.com/ZappoMan/hifi/6daf68b338282ad68f615339bc185c48a38d64ce/script-archive/controllers/puppetFeet3.js) and verify that the avatars left foot will "puppet" around from your left hand. It should be approximately 1 meter below your hand, and rotated from your hand rotation, but otherwise track your hand. (note: this is not a QA fail, if the IK doesn't always move the foot to exactly this position).

TODO:
- verify that with the post transform filter, the velocity is correct
- implement a test that shows this "child of the tracker" style filter

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: bab7d1e596ac5975dc73501a59033fa7c450d4fc available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-bab7d1e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-bab7d1e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6daf68b338282ad68f615339bc185c48a38d64ce available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-6daf68b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-6daf68b.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: afe7c386c3831ddc6e75d456fad1879a20f9a818 available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-afe7c38.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-afe7c38.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: afe7c386c3831ddc6e75d456fad1879a20f9a818 available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-afe7c38.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-afe7c38.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dce73ea428dfe7bbdebd05aed68ed2c4c332f79f available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-dce73ea.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-dce73ea.dmg

File: libraries/controllers/src/controllers/Pose.cpp [hunk 1]
```diff
@@ -69,5 +69,24 @@ namespace controller {
...
         return pose;
     }
+
+    Pose Pose::postTransform(const glm::mat4& mat) const {
+        glm::mat4 original = getMat4();
+        glm::mat4 result = original * mat;
+
+        auto translationOut = ::extractTranslation(result);
+        auto rotationOut = ::glmExtractRotation(result);
+        auto velocityOut = velocity + glm::cross(angularVelocity, translation - translationOut); // warning: this may be completely wrong
```
Comment (hyperlogic) [inline libraries/controllers/src/controllers/Pose.cpp]: Hey, this is close!  I think it should be (translationOut - translation).  Everything else looks good 👍

File: libraries/controllers/src/controllers/Pose.cpp [hunk 2]
```diff
@@ -69,5 +69,23 @@ namespace controller {
         pose.valid = valid;
         return pose;
     }
+
+    Pose Pose::postTransform(const glm::mat4& mat) const {
+        glm::mat4 original = ::createMatFromQuatAndPos(rotation, translation);
+        glm::mat4 result = original * mat;
+        auto translationOut = ::extractTranslation(result);
+        auto rotationOut = ::glmExtractRotation(result);
+        auto velocityOut = velocity + glm::cross(angularVelocity, translation - translationOut); // warning: this may be completely wrong
```
Comment (hyperlogic) [inline libraries/controllers/src/controllers/Pose.cpp]: This is very close! Except it should be (translationOut - translation), other then that it looks good

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8ff1ca8af6b933e2cf1be30b97a1f6360388df09 available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-8ff1ca8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-8ff1ca8.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 94addc0eec2d24c4e3f668bdd56e8bd8439a8a07 available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-94addc0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-94addc0.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7a1d91dbdd4182b9e3d03b7fd4e59da05d46c061 available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-7a1d91d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-7a1d91d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 829fa2b32357554ebbd9d5379d598db8336edb4a available at: https://gist.github.com/e543ec612e974a74297cbe5d07046944<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-829fa2b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10196-829fa2b.dmg

**Comment by Nex-Pro** [comment]
Checking this.

**Comment by Nex-Pro** [comment]
### Pass

This worked better after I hacked on the script a little. ( I got the foot to be vertically below the hand  )

No issues seen.

**Pull request merged**